### PR TITLE
feat: Implement support for KMS arguments

### PR DESCRIPTION
**Pull request by mkilchhofer**
## Describe your changes

This implements [Add Key Management Service (KMS) etcd encryption to an Azure Kubernetes Service (AKS) cluster](https://learn.microsoft.com/en-us/azure/aks/use-kms-etcd-encryption).

Feature is not yet available in azurerm provider, but also opened a PR over there:
- https://github.com/hashicorp/terraform-provider-azurerm/pull/19893

## Issue number

`-`

## Checklist before requesting a review
- [x] The pr title can be used to describe what this pr did in `CHANGELOG.md` file
- [x] I have executed pre-commit on my machine
- [x] I have passed pr-check on my machine

Thanks for your cooperation!

### CODE REVIEW

File: main.tf
Comment (lonegunmanb) [inline main.tf:140]: Since we could turn `enabled` to `false` meanwhile keeping this `key_vault_kms`, I would recommend the following toggle expression:

```hcl
for_each = var.key_vault_kms_enabled != null ? ["key_vault_kms"] : []
```

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @mkilchhofer for opening this pr and submitting a new resource pr to terraform-azurerm-provider's repo, I've left some comments on provider's pr, once the pr has been merged into the provider, we can continue the code review for this pr.


Btw the review comments for your provider's pr are my personal suggestions, there's no guarantee that your pr would be accepted by HashiCorp if you accepted my suggestion.

File: main.tf
Comment (mkilchhofer) [reply main.tf]: refactored this anyways (source was the upstream implemantation)

**Comment by mkilchhofer** [comment]
PR is now updated after the upstream provider PR got merged AND released ;-)

@lonegunmanb can you please review again?

File: variables.tf
Comment (lonegunmanb) [inline variables.tf:341]: Would an extra `nullable = false` here be better?

File: variables.tf
Comment (lonegunmanb) [inline variables.tf:353]: Should the default value be `Public` here?

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @mkilchhofer for updating this pr, some issues need to be solved.

It would be nice if we could have an new example to demonstrate this new feature. We have already created a KeyValut in `startup` example, but for disk encryption key management for now, would you please estimate whether if we can make that KeyVault work for this new feature? We might need to update `azurerm` provider's restriction in `providers.tf` file in `startup` folder. Or we can just crate a new example folder to do so, your call.

Again, thanks for your contribution!

File: variables.tf
Comment (mkilchhofer) [reply variables.tf:353]: Not sure if it is good practice to bubble defaults from the provider up to the module.
It's already bubbled up from go-sdk to terraform provider. So if it's set to `null`, it defaults to `Public` on the provider level.

But I am also fine do default it also here. :-)

File: variables.tf
Comment (mkilchhofer) [reply variables.tf:341]: I can do this, sure. 👍

File: main.tf
Comment (zioproto) [inline main.tf:163]: KMS etcd encryption doesn't work with system-assigned managed identity. You should check for `var.kms_enabled` and for `var.identity_type` to be `UserAssigned`.

**Review by zioproto — CHANGES_REQUESTED** [review]

File: variables.tf
Comment (mkilchhofer) [reply variables.tf:341]: Done

File: main.tf
Comment (mkilchhofer) [reply main.tf:163]: Done

**Comment by mkilchhofer** [comment]
I cannot test the examples as we @swisspost have no subscriptions which allows usage of public IPs.
Can someone test that for us?

**Comment by zioproto** [comment]
@mkilchhofer I can do that ! thanks

**Comment by zioproto** [comment]
@mkilchhofer  could you please run the `pre-commit` step on your local machine ?:

```
docker run --rm -v $(pwd):/src -w /src mcr.microsoft.com/azterraform:latest make pre-commit
```

You patched the example and you have duplicated definitions for `identity_type` and `identity_ids`. See the error:

```
Error: Attribute redefined
│
│   on examples/named_cluster/main.tf line 88, in module "aks_cluster_name":
│   88:   identity_type         = "UserAssigned"
│
│ The argument "identity_type" was already set at
│ examples/named_cluster/main.tf:66,3-16. Each argument may be set only once.
╵

╷
│ Error: Attribute redefined
│
│   on examples/named_cluster/main.tf line 89, in module "aks_cluster_name":
│   89:   identity_ids          = [azurerm_user_assigned_identity.test.id]
│
│ The argument "identity_ids" was already set at
│ examples/named_cluster/main.tf:65,3-15. Each argument may be set only once.
```

File: examples/named_cluster/main.tf
Comment (zioproto) [inline examples/named_cluster/main.tf:82]: These variable name is invalid. Should be just `kms_enabled`

File: examples/named_cluster/main.tf
Comment (zioproto) [inline examples/named_cluster/main.tf:83]: This should be `kms_key_vault_key_id` according to your change in `variable.tf`

**Review by zioproto — CHANGES_REQUESTED** [review]
Could you patch only the `named_cluster` example ? 

We dont need to the KMS in the others, it would also require more work because the needed `UserAssigned` identity_type is not set in the others. Thanks

File: examples/named_cluster/main.tf
Comment (mkilchhofer) [reply examples/named_cluster/main.tf:82]: Yea. Sorry, copy pasta error before refactoring :D

File: examples/named_cluster/main.tf
Comment (mkilchhofer) [reply examples/named_cluster/main.tf:83]: Also copy pasta error before refacotoring :/

**Comment by zioproto** [comment]
commit 70121785da3e7b0a78393f44b0c1a9c87566902b is now passing both `pre-commit` and `pr-check` steps.

@lonegunmanb could you please review and run the e2e tests ? thanks

File: .checkov_config.yaml
Comment (zioproto) [inline .checkov_config.yaml:14]: I believe this is OK because not all customers require the use of HSM for Azure Key Vault

File: examples/named_cluster/kms.tf
Comment (lonegunmanb) [inline examples/named_cluster/kms.tf:23]: We've got an e2e test error:

```text
TestExamplesNamedCluster 2023-01-27T04:53:34Z logger.go:66: │ Error: expected "object_id" to be a valid UUID, got /subscriptions/f7a632a5-49db-4c5e-9828-cd62cb753971/resourceGroups/8df3563089d5cbb3-rg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/8df3563089d5cbb3-identity
TestExamplesNamedCluster 2023-01-27T04:53:34Z logger.go:66: │ 
TestExamplesNamedCluster 2023-01-27T04:53:34Z logger.go:66: │   with azurerm_key_vault_access_policy.kms,
TestExamplesNamedCluster 2023-01-27T04:53:34Z logger.go:66: │   on kms.tf line 23, in resource "azurerm_key_vault_access_policy" "kms":
TestExamplesNamedCluster 2023-01-27T04:53:34Z logger.go:66: │   23:   object_id    = azurerm_user_assigned_identity.test.id
```

Could you please try `object_id = azurerm_user_assigned_identity.test.principal_id`, then try to pass `TestExamplesNamedCluster` test inside `test/e2e` folder? Thanks!

**Review by lonegunmanb — CHANGES_REQUESTED** [review]

File: variables.tf
Comment (mkilchhofer) [reply variables.tf:354]: 👍 will do

File: examples/named_cluster/kms.tf
Comment (mkilchhofer) [reply examples/named_cluster/kms.tf:23]: Cannot run E2E here as I have no access to a unrestricted Azure Subscription.

But jup, `principal_id` will fix it halfway, we also need a `depends_on` inside the call of the module to ensure the managed identity has the correct access policy applied before we call the managedCluster API.

Comment (mkilchhofer) [reply examples/named_cluster/kms.tf]: done

**Comment by mkilchhofer** [comment]
@lonegunmanb can you trigger E2E again? 😇

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @mkilchhofer for this update, we've met some issues when we ran the test.

We've added `network_acl` on the KeyVault to restrict the source IP which can access this Key Vault, but the Aks's public IP is unpredictable so the original code would meet a 403 error.

We need config the aks to access the KeyVault via private network as [this document](https://learn.microsoft.com/en-us/azure/aks/use-kms-etcd-encryption#create-an-aks-cluster-with-private-key-vault-and-enable-kms-etcd-encryption) described.

File: examples/named_cluster/main.tf
Comment (mkilchhofer) [reply examples/named_cluster/main.tf:80]: Service endpoint is not needed (at least in our scenario). I think its only needed if some part inside the VNET wants to access the keyvault.

File: examples/named_cluster/main.tf
Comment (mkilchhofer) [reply examples/named_cluster/main.tf:84]: We at @swisspost use private connection. Should the examples also use the private one?
Thought that `bypass         = "AzureServices"` will allow AKS to communicate with vault but it seems that its not covered in that.

**Review by lonegunmanb — APPROVED** [review]
Thanks @mkilchhofer, LGTM! 🚀

**Comment by mkilchhofer** [comment]
Thanks alot for your patience with me 👍

**Pull request merged**
